### PR TITLE
liveness "depth charge" health check

### PR DESCRIPTION
**Pull request by suchapalaver**
# [CHRON-344](https://blockchaintp.atlassian.net/browse/CHRON-344)

- [x] cli arg under `serve-api`- `--liveness-check`, if present without a value defaults to `"1800"`. For checks every five seconds: `serve-api --liveness-check 5`
- [x] task in api creation sending a define Activity transaction at an interval determined at runtime
- [x] task installs a Prometheus exporter - scrape endpoint on `[http://127.0.0.1:9000/metrics](http://127.0.0.1:9000/metrics)` - emitting histogram of elapsed time for depth charge round trip. [Plans are afoot](https://blockchaintp.atlassian.net/browse/CHRON-450) to make this a configurable endpoint.   
- [x] add Helm Chart support (`kubectl port-forward test-chronicle-chronicle-0 9000:9000`)

```yaml
---
imagePullSecrets:
  enabled: true
  value: 
    - name: regcred

healthCheck:
  enabled: true
  # To have it run every 5 seconds
  interval: 5

image: 
  pullPolicy: Always
  repository: dev.catenasys.com:8083/joseph/chronicle-amd64
  tag: latest
  
logLevel: trace,cranelift_codegen=off,wasmtime_cranelift=off,regalloc2::ion=off,cranelift_wasm=off
```

- [x] add documentation

# PR Checklist

## Please complete this checklist after opening your PR

* [x] I have updated documentation as necessary
* [x] I have updated helm chart(s) if needed
* [x] I have added tests if needed
* [ ] All new and existing tests are passing
* [x] Any breaking changes are clearly flagged and documented
* [x] I’ve included a link to any relevant ticket(s)

[CHRON-344]: https://blockchaintp.atlassian.net/browse/CHRON-344?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: crates/api/src/lib.rs [hunk 1]
```diff
@@ -266,11 +300,13 @@ where
             .run(|connection| connection.run_pending_migrations(MIGRATIONS).map(|_| ()))
             .map_err(|migration| StoreError::DbMigration(migration))?;
 
-        // Append namespace bindings and system namespace
-        store.namespace_binding(
+        let system_namespace_uuid = (
```
Comment (ryanbtp) [inline crates/api/src/lib.rs:303]: We should probably make this a static / const somewhere useful

File: crates/api/src/lib.rs [hunk 2]
```diff
@@ -374,6 +410,95 @@ where
...
+                                            error!("Depth charge transaction rejected by ledger: {id} {contradiction}");
+                                            break;
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        Err(e) => error!("ApiError submitting depth charge: {}", e),
+                        // No other ApiCommand types are handled by this loop
+                        _ => unreachable!(),
```
Comment (ryanbtp) [inline crates/api/src/lib.rs:496]: I would make this explicit Ok(response) -> error!()

**Review by ryanbtp — CHANGES_REQUESTED** [review]

File: docs/helm-options.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# Other Chronicle Helm Options
+
+## Liveness Health Check
+
+Chronicle can enable a liveness depth charge checks to ensure system availability.
```
Comment (mtbc) [inline docs/helm-options.md:5]:
```suggestion
Chronicle can enable a liveness depth charge check to ensure system availability.
```

**Comment by mtbc** [comment]
Could be a followup PR but: do we want this instead listening on `API_LISTEN_SOCKET` at `/metrics` or suchlike, using the handle from `install_recorder()`? Or okay for it to just use its defaults?

File: docs/health-checks-and-testing.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+[System Namespace](./namespaces.md#chronicle-system). At each interval,
+the depth charge operation is executed and timed. If the transaction is
+successfully committed, the elapsed time of the round trip is calculated
+and made available as a histogram metric.
+
+The liveness check task installs a [Prometheus](https://prometheus.io/)
+exporter to record metrics. If the transaction is successfully committed,
+the elapsed time of the round trip is calculated and recorded as a
+histogram metric in the Prometheus exporter. The collected data can be
+scraped from the exposed endpoint at `127.0.0.1:9000` for analysis and
```
Comment (mtbc) [inline docs/health-checks-and-testing.md:21]: It seems to be listening on every interface when I try it.

File: crates/api/src/lib.rs [hunk 5]
```diff
@@ -374,6 +407,94 @@ where
...
+            debug!("Starting liveness depth charge task");
+            let depth_charge_api = dispatch.clone();
+            tokio::task::spawn(async move {
+                if let Err(e) = PrometheusBuilder::new().install() {
+                    warn!(
+                        "Prometheus exporter installation for liveness check metrics failed: {:?}",
+                        e
+                    );
+                } else {
+                    debug!("Liveness check metrics Prometheus exporter for installed with endpoint on 127.0.0.1:9000");
```
[the remark this answers is not in the document]
Comment (suchapalaver) [reply crates/api/src/lib.rs]: Yes.

File: docs/helm-options.md (hunk 3 above)
Comment (suchapalaver) [reply docs/helm-options.md]: Thanks!

File: docs/health-checks-and-testing.md (hunk 4 above)
Comment (suchapalaver) [reply docs/health-checks-and-testing.md:21]: Thanks. We're at least no longer using the default initialization for the Prometheus builder that uses `0.0.0.0`.

**Comment by suchapalaver** [comment]
> Could be a followup PR but: do we want this instead listening on `API_LISTEN_SOCKET` at `/metrics` or suchlike, perhaps using the handle from `install_recorder()`? Or okay for it to just use its defaults?

I've refactored the code that initializes the recorder to a separate `fn` in `api/src/lib.rs` for now.

File: docs/helm-options.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+Chronicle can enable liveness depth charge checks to ensure system availability.
+
+The following options in the Chronicle Helm Chart `values.yaml` file can be used
+to enable and configure the health check:
+
+```yaml
+healthCheck:
+  enabled: false
+  interval:
```
Comment (mtbc) [inline docs/helm-options.md:13]:
```suggestion
  enabled: true
  interval: 1800
```
might be a helpful example.

**Comment by mtbc** [comment]
> improvements/requirements for subsequent work, perhaps in another ticket

Good plan, this PR is just about ready to go, have created CHRON-450 to note that outstanding thought.

**Comment by ryanbtp** [comment]
Yes we can leave the simple /health endpoint for another ticket - that can be metric based

**Review by ryanbtp — APPROVED** [review]

**Pull request merged**